### PR TITLE
Generative VQ-VAE model for OASIS dataset

### PR DESCRIPTION
**Pull request by sophie-bates**
Hi Shakes, 

This pull request contains my implementation of a generative model of the OASIS brain dataset that uses a DCGAN for novel image generation, and VQ-VAE for reconstructions. It has successfully achieved a Structured Similarity (SSIM) score greater than 0.6. 

All details required for understanding the implementation and reproducing results are contained in the README.md file. The key files included in this PR are as follows:

| **Module** | **Description** |
|---|---|
| [dataset.py](dataset.py) | Loads the data and preprocesses it for use by the train loaders. |
| [modules.py](modules.py) | Core components of the model required for the pattern recognition task, includes VQVAE (encoder, vector_quantizer, and decoder) and GAN models (discriminator, generator). |
| [predict.py](predict.py) | Example usage of the trained model, generates results and provides visualisations on the testing dataset. |
| [train.py](train.py) | Training script for the VQVAE and GAN models, including validation and saving of the model, and plotting losses and metrics observed during training. |

Looking forward to receiving feedback,

Sophie Bates
s4583766

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme Overall:  Good
- Project Overview: Good
- Model: Good
- Data + Preprocessing: Good
- Usage: Good
- Training / Loss Curve(s): Good
- Result Demonstration: Good
- References: Good

Functionality:
- Reconstruction: Good
- Generation: Not quite working but should be close.
- Note: The generator should probably predict codebook indices (class) using softmax. if codebook size is 512 entries then generator can predict bs x 512 x latent_h x latent_d. Apply softmax to get the indices then decode using the codebook + decoder. Tanh is probably not going to work.

Code:
- Consistent with Results: Yes
- File Structure: Good
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good

Other Comments:
N/A

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
